client/go/cmd/clone.go

```diff
@@ -108,71 +116,147 @@ func (c *cloner) Clone(applicationName, path string) error {
+	// the cached copy if GitHub is unavailable.
+	cacheHit := zipPath != ""
+	err := c.cli.spinner(c.cli.Stderr, color.YellowString("Downloading sample apps ..."), func() error {
 		request, err := http.NewRequest("GET", "https://github.com/vespa-engine/sample-apps/archive/refs/heads/master.zip", nil)
 		if err != nil {
 			return fmt.Errorf("invalid url: %w", err)
 		}
+		if etag != "" {
+			request.Header = make(http.Header)
+			request.Header.Set("if-none-match", fmt.Sprintf(`W/"%s"`, etag))
+		}
 		response, err := c.cli.httpClient.Do(request, time.Minute*60)
 		if err != nil {
 			return fmt.Errorf("could not download sample apps: %w", err)
 		}
 		defer response.Body.Close()
+		if response.StatusCode == http.StatusNotModified { // entity tag matched so our cached copy is current
+			return nil
+		}
 		if response.StatusCode != http.StatusOK {
 			return fmt.Errorf("could not download sample apps: github returned status %d", response.StatusCode)
 		}
-		if _, err := io.Copy(f, response.Body); err != nil {
-			return fmt.Errorf("could not write sample apps to file: %s: %w", f.Name(), err)
-		}
-		f.Close()
-		if err := os.Rename(f.Name(), dst); err != nil {
-			return fmt.Errorf("could not move sample apps to cache path")
+		etag = trimEntityTagID(response.Header.Get("etag"))
+		newPath, err := c.writeZip(response.Body, etag)
+		if err != nil {
+			return err
 		}
+		zipPath = newPath
+		cacheHit = false
 		return nil
 	})
+	return zipPath, cacheHit, err
 }
 
-func (c *cloner) zipName() (string, error) {
-	cacheDir, err := vespaCliCacheDir(c.cli.Environment)
+// writeZip atomically writes the contents of reader zipReader to a file in the CLI cache directory.
+func (c *cloner) writeZip(zipReader io.Reader, etag string) (string, error) {
+	f, err := os.CreateTemp(c.cli.config.cacheDir, "sample-apps-tmp-")
 	if err != nil {
-		return "", err
-	}
-	dst := filepath.Join(cacheDir, "sample-apps-master.zip")
-	cacheExists := true
-	stat, err := os.Stat(dst)
-	if errors.Is(err, os.ErrNotExist) {
-		cacheExists = false
-	} else if err != nil {
-		return "", fmt.Errorf("could not stat existing cache file: %w", err)
-	}
-	if cacheExists {
-		useCache, err := c.useCache(stat)
-		if err != nil {
-			return "", errHint(fmt.Errorf("could not determine cache status: %w", err), "Try ignoring the cache with the -f flag")
-		}
-		if useCache {
-			log.Print(color.YellowString("Using cached sample apps ..."))
-			return dst, nil
+		return "", fmt.Errorf("could not create temporary file: %w", err)
+	}
+	cleanTemp := true
+	defer func() {
+		f.Close()
+		if cleanTemp {
+			os.Remove(f.Name())
 		}
+	}()
+	if _, err := io.Copy(f, zipReader); err != nil {
+		return "", fmt.Errorf("could not write sample apps to file: %s: %w", f.Name(), err)
+	}
+	f.Close()
+	path := filepath.Join(c.cli.config.cacheDir, sampleAppsNamePrefix)
+	if etag != "" {
+		path += "_" + etag
```

Comment:
I am not entirely sold on this yet.
Won't be be less bookkeeping to keep one file for the content, and one for the etag. Then we don't need the directory traversal and garbage collection.

And if we default the etag to empty string for all etag file failure cases I think the rest should work itself out, even if we end up downloading the file a few more times than required.

Reply:
This way we guarantee that the data and its etag are in sync by not keeping state in a separate file. For example it's not possible to write the cached data, but not the etag, or vice-versa.

We also don't need special handling related to presence of an etag file, atomic writing of the etag file and parsing of its contents.